### PR TITLE
Support for mailers

### PR DESCRIPTION
**Pull request by paganotoni**
Hi @markbates i did some progress today on the mailer support, i'm adding this PR looking for your feedback on the things i've done.

Please let me know if i'm going through a correct path (or the opposite) on this once you have the chance, my thinking is i still need to support the correct encoding based on the content type as well as testing stronger the SMTP part for files.

### CODE REVIEW

**Comment by paganotoni** [comment]
I've been also thinking to extract the fake-SMTP part into its own lib but i don't want that to hold our progress here.

File: fakesmtp/fakesmtp.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		if bytes[0] == 46 && bytes[1] == 13 && bytes[2] == 10 {
+			break
+		}
+	}
+	c.write("250 server has transmitted the message")
+	c.conn.Close()
+}
+
+//Start listens for connections on the given port
+func WaitForMessage(port string) {
```
Comment (markbates) [inline fakesmtp/fakesmtp.go]: This doesn't actually "wait". :) You may want to not use a goroutine here, but instead let it block. That way, people can decide whether to let it block or not. 

I would also probably change it back to `Start` and have it return an error.

File: fakesmtp/fakesmtp.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"time"
+)
+
+var LastMessage string
```
Comment (markbates) [inline fakesmtp/fakesmtp.go]: I would probably get rid of this and have a `Messages()` method that returns the messages from the client, as well as a `Clear()` function that clears out all the saved messages between tests.

Comment (paganotoni) [reply fakesmtp/fakesmtp.go]: Sounds good, thanks Mark!

File: mailer/message.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+//Attachment are files added into a email message.
+type Attachment struct {
+	Name        string
+	Reader      io.Reader
+	ContentType string
+}
+
+// AddBody the message by receiving a renderer and rendering data.
+func (m *Message) AddBody(r render.Renderer, data render.Data, contentType string) error {
```
Comment (markbates) [inline mailer/message.go]: `render.Renderer` already has a `ContentType()` method as part of the interface, do we need the param `contentType` as well?

Comment (paganotoni) [reply mailer/message.go]: Awesome, didn't know about that one, thanks

**Comment by markbates** [comment]
@apaganobeleno one we merge https://github.com/gobuffalo/buffalo/pull/496 you'll be able to use it to write a generator for the mailers

**Comment by paganotoni** [comment]
@markbates i updated this one based on the review you made, thanks for that again, i think this one is ready to be reviewed, i used it today and seems to be working nice, there is one thing i've not tested (manually) and i will need a bit more of guidance on which i'll try to describe next:

On the `Deliver` function of the SMTPMailer (https://github.com/gobuffalo/x/pull/1/files#diff-3f10acccbe8f973df02d18cb41f4abedR26) we're doing  :

```go
m.SetBody(message.ContentType, string(message.Body), gomail.SetPartEncoding(gomail.Unencoded))
```

(Notice the gomail.Unencoded) section, on the gomail package there are 3 options (see [here](https://godoc.org/gopkg.in/gomail.v2#Encoding)) and i'm not sure if thats the correct one or it should be determined based on the message encoding.

Do you know where i can read/learn a bit more on this ?

**Comment by markbates** [comment]
Overall it looks good. I found one decent sized bug:

```go
func NewContactMessage(c *models.Contact) (mailer.Message, error) {
	m := mailer.NewMessage()
	m.From = "info@gopherguides.com"
	m.To = []string{"info@gopherguides.com"}
	data := map[string]interface{}{
		"contact": c,
	}
	err := m.AddBody(r.HTML("mail/new_contact.md"), data)
	if err != nil {
		return m, errors.WithStack(err)
	}
	err = m.AddBody(r.Template("text/plain", "mail/new_contact.md"), data)
	if err != nil {
		return m, errors.WithStack(err)
	}
	m.Subject = "New Contact | Gopher Guides"
	return m, nil
}
```

It appears that `AddBody` overwrites the previously set body. You should be able to, and it's quite common to, set multiple body types, for example a `text/plain` and a `text/html` body.

**Comment by markbates** [comment]
Also, please make sure to put a comment at the top of the fakesmtp script to attribute the original gist, I want to make sure we give credit where credit is due.

**Comment by paganotoni** [comment]
@markbates i'll check that bug shortly, also add that comment at the top of fakehttp, thanks for pointing that out.

**Comment by markbates** [comment]
Also, for your consideration, here is how I'm using it in a project:

```go
package mailers

import (
	"log"

	"github.com/gobuffalo/buffalo"
	"github.com/gobuffalo/buffalo/render"
	"github.com/gobuffalo/envy"
	"github.com/gobuffalo/packr"
	"github.com/gobuffalo/x/mailer"
	"github.com/gopherguides/web/models"
	"github.com/pkg/errors"
)

var smtp mailer.Deliverer
var r *render.Engine

func init() {
	var err error
	smtp, err = mailer.NewSMTPMailer(envy.Get("SMTP_PORT", "1025"), envy.Get("SMTP_HOST", "localhost"), envy.Get("SMTP_USER", ""), envy.Get("SMTP_PASSWORD", ""))
	if err != nil {
		log.Fatal(err)
	}
	r = render.New(render.Options{
		TemplatesBox: packr.NewBox("../templates"),
	})
}

type Message struct {
	mailer.Message
	smtp mailer.Deliverer
}

func NewMessage() Message {
	m := Message{
		Message: mailer.NewMessage(),
		smtp:    smtp,
	}

	m.From = "info@gopherguides.com"
	m.To = []string{"info@gopherguides.com"}
	return m
}

func (m Message) AddBodies(data map[string]interface{}, rs ...render.Renderer) error {
	for _, s := range rs {
		err := m.AddBody(s, data)
		if err != nil {
			return errors.WithStack(err)
		}
	}
	return nil
}

func (m Message) Deliver() error {
	return m.smtp.Deliver(m.Message)
}

func (m Message) DeliverLater(l buffalo.Logger) {
	go func() {
		err := m.smtp.Deliver(m.Message)
		if err != nil {
			l.Error(err)
		}
	}()
}

func NewContactMessage(c *models.Contact) (Message, error) {
	m := NewMessage()
	m.Subject = "New Contact | Gopher Guides"
	data := map[string]interface{}{
		"contact": c,
	}
	err := m.AddBodies(data, r.HTML("mail/new_contact.md"), r.Plain("mail/new_contact.md"))
	if err != nil {
		return m, errors.WithStack(err)
	}
	return m, nil
}

func NewTrainerOpportunityMessage(c *models.TrainerOpportunity) (Message, error) {
	m := NewMessage()
	m.Subject = "New Trainer Opportunity | Gopher Guides"
	data := map[string]interface{}{
		"trainerOpportunity": c,
	}
	err := m.AddBodies(data, r.HTML("mail/new_trainer_opportunity.md"), r.Plain("mail/new_trainer_opportunity.md"))
	if err != nil {
		return m, errors.WithStack(err)
	}
	return m, nil
}
```

The question is; is there anything we can take from this usage that would make this package better? I'm not sure, but just throwing it out there.

We also want to consider a generator at some point.

File: mailer/smtp_mailer.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+//NewSMTPMailer builds a SMTP Mailer based in passed config.
+func NewSMTPMailer(port string, host string, user string, password string) (SMTPMailer, error) {
```
Comment (markbates) [inline mailer/smtp_mailer.go]: It feels weird to have a port be the first argument. I feel like it should be `host`, `port`, `user`, `pass`.

**Comment by paganotoni** [comment]
@markbates i think that addBodies part could be added into the message, ~i bet you should have implemented it locally.~ i see you implemented it on your message type :D

I was also thinking on that "DeliverLater" functionality which we could add into the deliverer interface, but looks to me more related with worker/app-specific behaviour than the deliverer itself.

**Comment by paganotoni** [comment]
@markbates should we change the name of the SMTPMailer to be SMTPDeliverer ? that way we could save the "Mailer" suffix for the functions we're going to generating where developers would be using SMTPDeliverer as well as other Deliverers (i feel like talking about Justin Biever here lol).

File: fakesmtp/server.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
+package fakesmtp
+
+// This server is inspired by https://github.com/andrewarrow/jungle_smtp
+// and most of its functionallities have been taken from the original repo and updated to
```
Comment (markbates) [inline fakesmtp/server.go]: "functionality"

File: fakesmtp/server.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+//Clear the server messages
+func (s *Server) Clear() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.messages = []string{}
+}
+
+//NewServer returns a pointer to a new Server instance listening on the given port.
+func NewServer(port string) (*Server, error) {
```
Comment (markbates) [inline fakesmtp/server.go]: I think this can be just named `New`

**Comment by markbates** [comment]
I find the name `Deliverer` to difficult to say. I wonder if there is a better name. Thoughts? Related I find `SMTPDeliverer` clunky as well. Not sure of a better name for either of them though. :(

**Comment by paganotoni** [comment]
What about Sender ? we could change the interface to be named Sender and its function to be Send, does that sound better ?

**Comment by markbates** [comment]
`Sender` is better, not great, but better. Let's go with that.

File: fakesmtp/server.go (hunk 6 above)
Comment (paganotoni) [reply fakesmtp/server.go]: Awesome, changed

File: fakesmtp/server.go (hunk 5 above)
Comment (paganotoni) [reply fakesmtp/server.go]: Thanks, fixed.

**Comment by paganotoni** [comment]
I did a quick look at Sender synonyms and found:

broker
dealer
exporter
operator
retailer
seller
shipper
shopkeeper
trader
trafficker
vendor
wholesaler
businessperson
consigner
handler
jobber
marketer
salesperson
storekeeper
tycoon
tradesperson

**Comment by paganotoni** [comment]
@markbates what else do we need to add here ? documentation on the README ? or have a README inside the mailer package ?

**Comment by markbates** [comment]
A README inside the mailer package with some examples would be great.

**Comment by paganotoni** [comment]
Going for it.

File: mailer/README.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	m.Subject = "New Contact"
+	m.To = []string{"contact@myapp.com"}
+
+	// Data that will be used inside the templates when rendering.
+	data := map[string]interface{}{
+		"contact": c,
+	}
+
+	// You can add multiple bodies to the message you're creating to have content-types alternatives.
+	err := m.AddBodies(data, r.HTML("mail/contact.html"), r.Plain("mail/contact.txt"))
```
Comment (markbates) [inline mailer/README.md:48]: Where does `r` come from in this example? We should show setting it up so people don't just cut and paste and can't figure out what it comes from.

**Comment by markbates** [comment]
@apaganobeleno I merged this because I want to use, but if you could update that README it would be great.

**Pull request merged**